index.js

```diff
@@ -2,271 +2,169 @@
-const BENDYBT_MSG_TF = Buffer.from([MSG.code, MSG.formats['ssb/bendy-butt'].code])
-
-const BLOB = NAMED_TYPES['blob']
-const BLOB_T = Buffer.from([BLOB.code])
-const CLASSIC_BLOB_TF = Buffer.from([BLOB.code, BLOB.formats['ssb/classic'].code])
-
-const SIGNATURE = NAMED_TYPES['signature']
-const SIGNATURE_TF = Buffer.from([SIGNATURE.code, SIGNATURE.formats['ed25519'].code])
-
-const BOX = NAMED_TYPES['encrypted']
-const BOX_T = Buffer.from([BOX.code])
-const BOX1_TF = Buffer.from([BOX.code, BOX.formats['box1'].code])
-const BOX2_TF = Buffer.from([BOX.code, BOX.formats['box2'].code])
+const TYPES = bufferizeCodes(raw)
+const NAMED_TYPES = sortByNames(TYPES)
 
 const GENERIC = NAMED_TYPES['generic']
-const STRING_TF = Buffer.from([GENERIC.code, GENERIC.formats['UTF8 string'].code])
-const BOOL_TF = Buffer.from([GENERIC.code, GENERIC.formats['boolean'].code])
+const STRING_TF = Buffer.concat([
+  GENERIC.code,
+  GENERIC.formats['UTF8 string'].code,
+])
+const BOOL_TF = Buffer.concat([GENERIC.code, GENERIC.formats['boolean'].code])
 const BOOL_TRUE = Buffer.from([1])
 const BOOL_FALSE = Buffer.from([0])
-const NIL_TF = Buffer.from([GENERIC.code, GENERIC.formats['nil'].code])
+const NIL_TF = Buffer.concat([GENERIC.code, GENERIC.formats['nil'].code])
 const NIL_TFD = NIL_TF
 
 const encoder = {
-  feed(feedId) {
-    let tf
-    if (feedId.endsWith('.ed25519')) tf = CLASSIC_FEED_TF
-    else if (feedId.endsWith('.ggfeed-v1')) tf = GABBYGR_FEED_TF
-    else if (feedId.endsWith('.bbfeed-v1')) tf = BENDYBT_FEED_TF
-    else throw new Error('Unknown feed format: ' + feedId)
-
-    const dotIndex = feedId.lastIndexOf('.')
-    const b64part = feedId.substring(1, dotIndex)
-    const d = Buffer.from(b64part, 'base64')
-
-    return Buffer.concat([tf, d])
+  uri(input) {
+    let [type, format, ...data] = input.slice(6).split('/')
+    // *FIRE* WARNING ... are we URIEncoding our URIs?!
+    type = NAMED_TYPES[type]
+    if (type) {
+      format = type.formats[format]
+      if (format) {
+        return Buffer.concat([
+          type.code,
+          format.code,
+          Buffer.from(data.join('/'), 'base64'),
+        ])
+      }
+    }
+    throw new Error(`unknown type/format ${type.type}/${format.format}`)
   },
-
-  message(msgId) {
-    let tf
-    if (msgId.endsWith('.sha256')) tf = CLASSIC_MSG_TF
-    else if (msgId.endsWith('.ggmsg-v1')) tf = GABBYGR_MSG_TF
-    else if (msgId.endsWith('.bbmsg-v1')) tf = BENDYBT_MSG_TF
-    else throw new Error('Unknown msg ID: ' + msgId)
-
-    const dotIndex = msgId.lastIndexOf('.')
-    const b64part = msgId.substring(1, dotIndex)
-    const d = Buffer.from(b64part, 'base64')
-
-    return Buffer.concat([tf, d])
+  boolean(input) {
+    const d = input ? BOOL_TRUE : BOOL_FALSE
+    return Buffer.concat([BOOL_TF, d])
   },
+  classic(input, type, format) {
+    let data = input
+    if (type.sigil) data = data.slice(1)
+    if (format.suffix) data = data.replace(format.suffix, '')
```

Comment:
`data.slice(0, -format.suffix.length)` would be safer here because we know for sure that the removal is happening at the end of the string. This code is probably safe, but not future-proof (who knows if there would be a suffix `foo` in the future which can also randomly occur in the middle of the base64 string).

Reply:
oh didn't know you could do this, that's awesome